x/meshsecurity/keeper/stake.go

```diff
@@ -58,12 +58,47 @@ func (k Keeper) Delegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.Val
 	// and update our records
 	k.setTotalDelegatedAmount(cacheCtx, actor, newTotalDelegatedAmount)
 	done()
-
-	// TODO: emit events?
-	// TODO: add to telemetry?
 	return newShares, err
 }
 
 func (k Keeper) Undelegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.ValAddress, amt sdk.Coin) error {
-	panic("not implemented, yet")
+	if amt.Amount.IsZero() || amt.Amount.IsNegative() {
+		return errors.ErrInvalidRequest.Wrap("amount")
+	}
+
+	// Ensure staking constraints
+	bondDenom := k.staking.BondDenom(pCtx)
+	if amt.Denom != bondDenom {
+		return errors.ErrInvalidRequest.Wrapf("invalid coin denomination: got %s, expected %s", amt.Denom, bondDenom)
+	}
+
+	cacheCtx, done := pCtx.CacheContext() // work in a cached store (safety net?)
+	shares, err := k.staking.ValidateUnbondAmount(cacheCtx, actor, valAddr, amt.Amount)
+	if err == stakingtypes.ErrNoDelegation {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	undelegatedCoins, err := k.staking.InstantUndelegate(cacheCtx, actor, valAddr, shares)
+	if err != nil {
+		return err
+	}
+	err = k.bank.SendCoinsFromAccountToModule(cacheCtx, actor, types.ModuleName, undelegatedCoins)
+	if err != nil {
+		return err
+	}
+
+	err = k.bank.BurnCoins(cacheCtx, types.ModuleName, undelegatedCoins)
+	if err != nil {
+		return err
+	}
+
+	unbondedAmount := undelegatedCoins.AmountOf(bondDenom)
+	k.bank.AddSupplyOffset(cacheCtx, bondDenom, unbondedAmount)
+	newDelegatedAmt := k.getTotalDelegatedAmount(cacheCtx, actor).Sub(unbondedAmount)
```

Comment:
I would assert this in the beginning.

1. The calling account has a max_cap at all
2. The calling account has delegated >= amt

And then the rest of the (expensive) staking calls

Reply:
Good thinking! 
I will add a note to the doc about authorization. This needs to be handled outside. A max cap reduced to `0` may trigger an instant undelegate by the contract.
The delegated amount is checked early within staking. There is no separate record for an individual validator in MS. We can check against the total used amount in delegates as this is loaded anyway. This would be more of a sanity check rather than an efficient constraint, which is in staking. 
I will add this as it is cheap and also some code to ensure the "total delegated amount" can not be negative.

Reply:
Okay, then a quick sanity check that `max_cap > 0` at the top.
No one else should be able to call this at all.

Reply:
I think `max_cap == 0` is still a valid scenario where the contract needs to be able to undelegate. 
This would be an exit scenario where the consumer chain decides to set the max cap to `0` via gov. The contract would act on a callback and run undelegates to burn the "virtual" tokens.
There is a constraint in place that not more "virtual" tokens can be burned than were created.
Alternatively, the chain would have to undelegate and burn which may make it more complicated with the contracts in place

Reply:
The only "danger" scenario here is a malicious contract that wouldn't process the new max_cap correctly.

But this contract must have been approved and reviewed by the consumer chain governance, so I guess the consumer chain should trust it (even if they don't 100% trust the provider chain).

Fair enough. When max_cap is changed, we can trigger the `SudoMsg::Rebalance` call the following EndBlock (rather than waiting up to 1 day), so that this change is reflected asap. And then leave it to the contract to properly instant unbond and burn.